File: Java/LeetCode.java
/*
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.
*/

class Solution {
    public int[] twoSum(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[j] == target - nums[i]) {
                    return new int[] { i, j };
                }
            }
        }
        throw new IllegalArgumentException("No two sum solution");
    }
}

//Given an array nums and a value val, remove all instances of that value in-place and return the new length.
    
class Solution {
    public int removeElement(int[] nums, int val) {
        int i = 0;
        for (int j = 0; j < nums.length; j++) {
            if (nums[j] != val) {
                nums[i] = nums[j];
                i++;
            }
        }
        return i;
    }
}

//Determine whether an integer is a palindrome. An integer is a palindrome when it reads the same backward as forward.

class Solution {
    public boolean isPalindrome(int x) {
        
        int rev = 0,val;
        val = x;
        while(x > 0) { 
            rev = rev * 10 + x % 10; 
            x = x / 10; 
        } 
        if(val==rev) return true;
        else return false;
        
    }
}

/*
Given two binary trees and imagine that when you put one of them to cover the other, 
some nodes of the two trees are overlapped while the others are not.
You need to merge them into a new binary tree. 
The merge rule is that if two nodes overlap, then sum node values up as the new value of the merged node. 
Otherwise, the NOT null node will be used as the node of new tree.

 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode mergeTrees(TreeNode t1, TreeNode t2) {
        if (t1 == null)
            return t2;
        if (t2 == null)
            return t1;
        t1.val += t2.val;
        t1.left = mergeTrees(t1.left, t2.left);
        t1.right = mergeTrees(t1.right, t2.right);
        return t1;
    }
}

/*
Given the root node of a binary search tree, return the sum of values of all nodes with value between L and R (inclusive).
The binary search tree is guaranteed to have unique values.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public int rangeSumBST(TreeNode root, int L, int R) {
        
        int ans = 0;
        Stack<TreeNode> st = new Stack<TreeNode>();
        if (root == null) return 0;
        st.push(root);
        while (st.size() > 0) {
            TreeNode p = st.peek();
            st.pop();
            if (p != null) {
                if (p.val >= L && p.val <= R) {
                    ans += p.val;   // update the sum
                }
                if (L < p.val) {
                    st.push(p.left);  // push left tree
                }
                if (p.val < R) {
                    st.push(p.right);  // push the right tree
                }
            }
        }
        return ans;  
    }
}

File: Python/LeetCode.py
#Implement function ToLowerCase() that has a string parameter str, and returns the same string in lowercase.

class Solution:
    def toLowerCase(self, str: str) -> str:
        return str.lower()

#Invert Binary Tree.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        
        if root is None:
            return root
        
        self.invertTree(root.left)
        self.invertTree(root.right)
        
        t = root.left
        root.left = root.right
        root.right = t
        
        return root

#Given an array A of distinct integers sorted in ascending order, return the smallest index i that satisfies A[i] == i.  
#Return -1 if no such i exists.    
    
class Solution:
    def fixedPoint(self, A: List[int]) -> int:
        length = len(A)
        for i in range(length): 
            if A[i] == i: 
                return i 
        return -1
    
Given a string, you need to reverse the order of characters in each word within a sentence while still preserving whitespace and initial word order.
    
class Solution:
    def reverseWords(self, s: str) -> str:
        def reverseWord(word):
            return word[::-1]
    
        s_split = s.split(' ')
        s_reverse = [reverseWord(i) for i in s_split]
        return ' '.join(s_reverse)
